numpy/core/src/multiarray/ctors.c

```diff
@@ -1809,37 +1813,19 @@ PyArray_FromStructInterface(PyObject *input)
  */
 NPY_NO_EXPORT int
 _is_default_descr(PyObject *descr, PyObject *typestr) {
-    PyObject *tuple, *name, *typestr2;
-    PyObject *tmp = NULL;
-    int ret = 0;
-
     if (!PyList_Check(descr) || PyList_GET_SIZE(descr) != 1) {
         return 0;
     }
-    tuple = PyList_GET_ITEM(descr, 0);
+    PyObject *tuple = PyList_GET_ITEM(descr, 0);
     if (!(PyTuple_Check(tuple) && PyTuple_GET_SIZE(tuple) == 2)) {
         return 0;
     }
-    name = PyTuple_GET_ITEM(tuple, 0);
+    PyObject *name = PyTuple_GET_ITEM(tuple, 0);
     if (!(PyUnicode_Check(name) && PyUnicode_GetLength(name) == 0)) {
         return 0;
     }
-    typestr2 = PyTuple_GET_ITEM(tuple, 1);
-    /* Allow unicode type strings */
-    if (PyUnicode_Check(typestr2)) {
-        tmp = PyUnicode_AsASCIIString(typestr2);
-        if (tmp == NULL) {
-            return 0;
-        }
-        typestr2 = tmp;
-    }
-    if (PyBytes_Check(typestr2) &&
-            PyObject_RichCompareBool(typestr, typestr2, Py_EQ)) {
-        ret = 1;
-    }
-    Py_XDECREF(tmp);
-
-    return ret;
+    PyObject *typestr2 = PyTuple_GET_ITEM(tuple, 1);
+    return PyObject_RichCompareBool(typestr, typestr2, Py_EQ);
```

Comment:
LGTM, sorry one last question to be sure.  We basically could get a different result now in such a case:
```
a = np.arange(5)
class iface:
    def __init__(self, a):
        iface = a.__array_interface__
        iface["typestr"] = iface["typestr"].encode()
        self.a = a
        self.__array_interface__ = iface

np.asarray(iface(a)).dtype  == a.dtype  # should not be structured
```
Where I think the change to byte string should make this comparison be false, leading to the use of `descr`, which will make it a structured result?

This is all super weird, and I recall you guys discussing some of the strange corners of the array-interface, so this may have been part of it..

Reply:
The line you're commenting on is for void types only I think, so should not affect that example.

I think the only change here is that mixing unicode an bytes in typestr vs descr no longer takes the fast path.

Reply:
But maybe I should test locally

Reply:
nvm, I just tested it locally and the example is unchanged. Seems I just did not dig deep enough trying to parse the code. Thanks!

Reply:
Here's the case you were thinking of:
```
In [34]: np.asarray(iface(np.zeros(1, 'V4'))).dtype
Out[34]: dtype([('f0', 'V4')])

In [35]: np.zeros(1, 'V4').dtype
Out[35]: dtype('V4')
```

So yes, this does change behavior, but if you're going to mix bytes and strings in the same object you probably deserve it